sw/device/lib/dif/dif_hmac.h

```diff
@@ -0,0 +1,289 @@
+                                         const void *data, size_t len,
+                                         size_t *bytes_sent);
+
+/**
+ * Retrieves the number of entries in the HMAC FIFO. These entries may be
+ * semi-arbitrary in length; this function should not be used to calculate
+ * message length.
+ *
+ * @param hmac The HMAC device to get the FIFO depth for.
+ * @param num_entries The number of entries in the FIFO.
+ * @return |kHmacErrorInvalidArgument| if |hmac| or |num_entries| is null,
+ *         |kHmacErrorNone| otherwise.
+ */
+dif_hmac_error_t dif_hmac_fifo_count_entries(const dif_hmac_t *hmac,
+                                             uint32_t *num_entries);
+
+/**
+ * Retrieves the number of bits in the loaded HMAC device.
+ * |dif_hmac_fifo_count_entries()| should be called before this function to
+ * ensure the FIFO is empty, as any bits in the FIFO are not counted in
+ * |msg_len|.
+ *
+ * @param hmac The HMAC device to get the message length for.
+ * @param msg_len The number of bits in the HMAC message.
+ * @return |kHmacErrorInvalidArgument| if |hmac| or |msg_len| is null,
+ * |kHmacErrorNone| otherwise.
+ */
+dif_hmac_error_t dif_hmac_get_message_length(const dif_hmac_t *hmac,
+                                             uint64_t *msg_len);
+
+/**
+ * Attempts to run HMAC or SHA256 depending on the mode |hmac| was initialized
+ * in. Calls to this function always succeed and return without blocking. The
+ * caller can use |dif_hmac_check_state()| to check for errors and for the
+ * |DIF_HMAC_DONE| status before reading the digest with
+ * |di_hmac_digest_read()|.
+ *
+ * @param hmac The HMAC device to initiate the run on.
+ * @return |kHmacErrorInvalidArgument| if |hmac| is null |kHmacErrorNone|
+ * otherwise.
+ */
+dif_hmac_error_t dif_hmac_run(dif_hmac_t *hmac);
+
+/**
+ * Checks the HMAC device to see if it has finished a run triggered by
+ * |dif_hmac_run()|. A logical true return value indicates that the HMAC run has
+ * completed and it is now safe to read the digest with
+ * |dif_hmac_digest_read()|.
+ *
+ * Calling this function has the side effect of clearing the hmac_done status,
+ * meaning that after this function returns true subsequent calls will return
+ * false until |dif_hmac_run()| is called again.
+ *
+ * @param hmac The HMAC device to check.
+ * @param done Out parameter, set to |true| if this is the first time this
+ *             function has been called since HMAC completed, |false| otherwise.
+ * @return |kHmacErrorInvalidArgument| if |hmac| is null |kHmacErrorNone|
+ * otherwise.
+ */
+dif_hmac_error_t dif_hmac_done(const dif_hmac_t *hmac, bool *done);
+
+/**
+ * Attempts to read the HMAC digest and store store the result in the buffer
+ * referenced by |digest_out|. |dif_hmac_state()| should before this
+ * function to ensure the digest is available to be read.
+ *
+ * |digest_out| must reference an allocated, contiguous, 32-byte buffer.
```

Comment:
I am by no means an expert on alignment. Am I correct in my belief that `dif_hmac_digest_t` must be 4-byte aligned since `uint_32` is its largest member, and that this is the only alignment requirement?

Reply:
You're approximately correct.

In C, arrays are aligned to the same alignment as their element type. C does not specify the alignment of structs, but various ABIs do, and in short structs get aligned to the alignment of their _first_ member (not their largest, I get caught out by this all the time). All scalars in C are aligned to at least their size (this is called "natural alignment").